mesonbuild/mupgrade.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2018 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys, os
+import json
+import subprocess
+import shlex
+from .mesonlib import meson_command
+from .compilers.compilers import compiler_envvars
+
+dump_format_version = 1
+
+def upgrade(state_file):
+    d = json.load(open(state_file, 'r'))
+    version = d.get('dump_format_version', 'missing')
+    if version != dump_format_version:
+        sys.exit('Can not upgrade because state dump file format is not compatible: %d != %d.' % (version, dump_format_version))
+    state = d['state']
+    env = os.environ.copy()
+    for c in state['compilers']:
+        ename = compiler_envvars[c[0]]
+        eval = ' '.join([shlex.quote(x) for x in c[1]])
+        env[ename] = eval
+    cmd_args = [state['source_root'],
+                state['build_root']]
+    if 'cross_file' in state:
+        cmd_args += ['--cross-file', state['cross_file']]
+    for o in state['options']:
+        cmd_args.append('-D%s=%s' % (o[0], o[1]))
```

Comment:
I think you need to add quotes because the value could contain spaces. Does it work with arrays? Also install_umask could be special here because it uses octal notation IIRC.